helper/activedirectory/client.go

```diff
@@ -0,0 +1,232 @@
+		Filter: toFilterString(filters),
+	}
+
+	conn, err := c.getFirstSucceedingConnection()
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	result, err := conn.Search(req)
+	if err != nil {
+		return nil, err
+	}
+
+	entries := make([]*Entry, len(result.Entries))
+	for i, rawEntry := range result.Entries {
+		entries[i] = NewEntry(c.logger, rawEntry)
+	}
+	return entries, nil
+}
+
+func (c *client) UpdateEntry(baseDNValues []string, filters map[*Field][]string, newValues map[*Field][]string) error {
+
+	entries, err := c.Search(baseDNValues, filters)
+	if err != nil {
+		return err
+	}
+	if len(entries) != 1 {
+		return fmt.Errorf("filter of %s doesn't match just one entry: %s", filters, entries)
+	}
+
+	replaceAttributes := make([]ldap.PartialAttribute, len(newValues))
+	i := 0
+	for field, vals := range newValues {
+		replaceAttributes[i] = ldap.PartialAttribute{
+			Type: field.String(),
+			Vals: vals,
+		}
+		i++
+	}
+
+	modifyReq := &ldap.ModifyRequest{
+		DN:                entries[0].DN,
+		ReplaceAttributes: replaceAttributes,
+	}
+
+	conn, err := c.getFirstSucceedingConnection()
+	if err != nil {
+		return err
+	}
+
+	return conn.Modify(modifyReq)
+}
+
+// UpdatePassword uses a Modify call under the hood because
+// Active Directory doesn't recognize the passwordModify method.
+// See https://github.com/go-ldap/ldap/issues/106
+// for more.
+func (c *client) UpdatePassword(baseDNValues []string, filters map[*Field][]string, newPassword string) error {
+
+	if !c.conf.StartTLS {
+		return errors.New("per Active Directory, a TLS session must be in progress to update passwords, please update your StartTLS setting")
+	}
+
+	pwdEncoded, err := formatPassword(newPassword)
+	if err != nil {
+		return err
+	}
+
+	newValues := map[*Field][]string{
+		FieldRegistry.UnicodePassword: {pwdEncoded},
+	}
+
+	return c.UpdateEntry(baseDNValues, filters, newValues)
+}
+
+func (c *client) getFirstSucceedingConnection() (ldapifc.Connection, error) {
+
+	var retErr *multierror.Error
+
+	tlsConfigs, err := c.conf.GetTLSConfigs()
+	if err != nil {
+		return nil, err
+	}
+
+	for u, tlsConfig := range tlsConfigs {
+		conn, err := c.connect(u, tlsConfig)
+		if err != nil {
+			retErr = multierror.Append(retErr, fmt.Errorf("error parsing url %v: %v", u, err.Error()))
+			continue
+		}
+
+		if c.conf.Username != "" && c.conf.Password != "" {
+			if err := conn.Bind(c.conf.Username, c.conf.Password); err != nil {
+				retErr = multierror.Append(retErr, fmt.Errorf("error binding to url %s: %s", u, err.Error()))
+				continue
+			}
+		}
+
+		return conn, nil
+	}
+
+	if c.logger.IsDebug() {
+		c.logger.Debug("ldap: errors connecting to some hosts: %s", retErr.Error())
+	}
+
+	return nil, retErr
+}
+
+func (c *client) connect(u *url.URL, tlsConfig *tls.Config) (ldapifc.Connection, error) {
+
+	_, port, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		// err intentionally ignored - we'll fall back to default ldap ports if we're unable to parse this
```

Comment:
Probably safer to assign port = "" here (which you use below). No guarantees are made what it will be if err != nil.